
Co-authored by Nali Toja

--- a/sys/conf/kern.mk
+++ b/sys/conf/kern.mk
@@ -3,6 +3,16 @@
 # Compat
 MK_FORMAT_EXTENSIONS?=no
 
+FREEBSD_CC!=   ${CC} -dM -E -</dev/null | awk '/__FreeBSD_cc_version/ { print $$3 }' || true
+
+.if ${FREEBSD_CC}
+# FreeBSD extensions
+format_extensions=     -fformat-extensions
+no_align_long_strings= -mno-align-long-strings
+.else
+MK_FORMAT_EXTENSIONS=no
+.endif
+
 #
 # Warning flags for compiling the kernel and components of the kernel:
 #
@@ -64,7 +74,7 @@
 #
 .if ${MACHINE_CPUARCH} == "i386"
 .if ${COMPILER_TYPE} != "clang"
-CFLAGS+=	-mno-align-long-strings -mpreferred-stack-boundary=2
+CFLAGS+=	${no_align_long_strings} -mpreferred-stack-boundary=2
 .else
 CFLAGS+=	-mno-aes -mno-avx
 .endif
