
Co-authored by Nali Toja

--- a/sys/conf/kern.mk
+++ b/sys/conf/kern.mk
@@ -64,6 +64,16 @@
 FORMAT_EXTENSIONS=	-fformat-extensions
 .endif
 
+FREEBSD_CC!=   ${CC} -dM -E -</dev/null | awk '/__FreeBSD_cc_version/ { print $$3 }' || true
+
+.if ${FREEBSD_CC}
+# FreeBSD extensions
+format_extensions=     -fformat-extensions
+no_align_long_strings= -mno-align-long-strings
+.else
+MK_FORMAT_EXTENSIONS=no
+.endif
+
 #
 # On i386, do not align the stack to 16-byte boundaries.  Otherwise GCC 2.95
 # and above adds code to the entry and exit point of every function to align the
@@ -84,7 +94,7 @@
 # Setting -mno-sse implies -mno-sse2, -mno-sse3, -mno-ssse3, -mno-sse41 and -mno-sse42
 #
 .if ${MACHINE_CPUARCH} == "i386"
-CFLAGS.gcc+=	-mno-align-long-strings -mpreferred-stack-boundary=2
+CFLAGS.gcc+=	${no-align-long-strings} -mpreferred-stack-boundary=2
 CFLAGS.clang+=	-mno-aes -mno-avx
 CFLAGS+=	-mno-mmx -mno-sse -msoft-float
 INLINE_LIMIT?=	8000
