Index: kfreebsd-10/sys/conf/newvers.sh
===================================================================
--- kfreebsd-10.orig/sys/conf/newvers.sh	2012-11-02 12:57:20.000000000 -0700
+++ kfreebsd-10/sys/conf/newvers.sh	2012-11-02 12:57:47.019211750 -0700
@@ -154,7 +154,7 @@
 cat << EOF > vers.c
 $COPYRIGHT
 #define SCCSSTR "@(#)${VERSION} #${v}${svn}${git}${p4version}: ${t}"
-#define VERSTR "${VERSION} #${v}${svn}${git}${p4version}: ${t}\\n    ${u}@${h}:${d}\\n"
+#define VERSTR "#${v} ${t}\\n"
 #define RELSTR "${RELEASE}"
 
 char sccs[sizeof(SCCSSTR) > 128 ? sizeof(SCCSSTR) : 128] = SCCSSTR;
