
freebsd-buildutils 9.0 doesn't define MK_CTF.

Index: kfreebsd-10/sys/conf/kern.post.mk
===================================================================
--- kfreebsd-10.orig/sys/conf/kern.post.mk	2012-11-02 12:57:46.469212125 -0700
+++ kfreebsd-10/sys/conf/kern.post.mk	2012-11-02 12:57:48.969213512 -0700
@@ -125,7 +125,7 @@
 	@rm -f ${.TARGET}
 	@echo linking ${.TARGET}
 	${SYSTEM_LD}
-.if ${MK_CTF} != "no"
+.if 0
 	${CTFMERGE} ${CTFFLAGS} -o ${.TARGET} ${SYSTEM_OBJS} vers.o
 .endif
 .if !defined(DEBUG)
Index: kfreebsd-10/sys/conf/kern.pre.mk
===================================================================
--- kfreebsd-10.orig/sys/conf/kern.pre.mk	2012-11-02 12:57:48.679211527 -0700
+++ kfreebsd-10/sys/conf/kern.pre.mk	2012-11-02 12:57:48.969213512 -0700
@@ -142,7 +142,7 @@
 ZFS_C=		${CC} -c ${ZFS_CFLAGS} ${WERROR} ${PROF} ${.IMPSRC}
 ZFS_S=		${CC} -c ${ZFS_ASM_CFLAGS} ${WERROR} ${.IMPSRC}
 
-.if ${MK_CTF} != "no"
+.if 0
 NORMAL_CTFCONVERT=	${CTFCONVERT} ${CTFFLAGS} ${.TARGET}
 .elif ${MAKE_VERSION} >= 5201111300
 NORMAL_CTFCONVERT=
