--- a/sys/amd64/conf/GENERIC
+++ b/sys/amd64/conf/GENERIC
@@ -18,8 +18,8 @@
 #
 # $FreeBSD$
 
-cpu		HAMMER
-ident		GENERIC
+#cpu		HAMMER
+#ident		GENERIC
 
 makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
 makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
@@ -53,7 +53,7 @@
 options 	COMPAT_FREEBSD5		# Compatible with FreeBSD5
 options 	COMPAT_FREEBSD6		# Compatible with FreeBSD6
 options 	COMPAT_FREEBSD7		# Compatible with FreeBSD7
-options 	SCSI_DELAY=5000		# Delay (in ms) before probing SCSI
+options 	SCSI_DELAY=2000		# Delay (in ms) before probing SCSI
 options 	KTRACE			# ktrace(1) support
 options 	STACK			# stack(9) support
 options 	SYSVSHM			# SYSV-style shared memory
--- a/sys/i386/conf/GENERIC
+++ b/sys/i386/conf/GENERIC
@@ -18,10 +18,10 @@
 #
 # $FreeBSD$
 
-cpu		I486_CPU
-cpu		I586_CPU
-cpu		I686_CPU
-ident		GENERIC
+#cpu		I486_CPU
+#cpu		I586_CPU
+#cpu		I686_CPU
+#ident		GENERIC
 
 makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
 makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
@@ -54,7 +54,7 @@
 options 	COMPAT_FREEBSD5		# Compatible with FreeBSD5
 options 	COMPAT_FREEBSD6		# Compatible with FreeBSD6
 options 	COMPAT_FREEBSD7		# Compatible with FreeBSD7
-options 	SCSI_DELAY=5000		# Delay (in ms) before probing SCSI
+options 	SCSI_DELAY=2000		# Delay (in ms) before probing SCSI
 options 	KTRACE			# ktrace(1) support
 options 	STACK			# stack(9) support
 options 	SYSVSHM			# SYSV-style shared memory
--- /dev/null
+++ b/sys/conf/DEBIAN
@@ -0,0 +1,36 @@
+# Common options to all Debian GNU/kFreeBSD kernels
+
+options		LINPROCFS
+options         LINSYSFS
+options		FDESCFS
+options		TMPFS
+options         QUOTA
+options		KTRACE
+
+# See #706736
+options		INCLUDE_CONFIG_FILE
+
+# Floppy drives need not be built-in because our installer doesn't
+# include a "floppy" flavour (and probably never will).
+nodevice	fdc
+
+# Network options
+device		carp
+device		pf
+device		pfsync
+device		pflog
+
+# Toggle to enable/disable debug options.
+include		DEBIAN_DEBUG
+#include	DEBIAN_NODEBUG
+
+# Alternate queueing
+options		ALTQ
+options		ALTQ_CBQ	# Class Bases Queuing (CBQ)
+options		ALTQ_RED	# Random Early Detection (RED)
+options		ALTQ_RIO	# RED In/Out
+options		ALTQ_HFSC	# Hierarchical Packet Scheduler (HFSC)
+options		ALTQ_PRIQ	# Priority Queuing (PRIQ)
+
+# Disable binary blobs
+include		WITHOUT_SOURCELESS
--- /dev/null
+++ b/sys/conf/DEBIAN_DEBUG
@@ -0,0 +1,20 @@
+makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
+
+makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
+
+## KDTRACE_FRAME is only for amd64
+#options		KDTRACE_FRAME		# Ensure frames are compiled in
+options		KDTRACE_HOOKS		# Kernel DTrace hooks
+options		DDB_CTF			# Kernel ELF linker loads CTF data
+
+options 	DDB			# Support DDB.
+options 	GDB			# Support remote GDB.
+options 	DEADLKRES		# Enable the deadlock resolver
+options 	INVARIANTS		# Enable calls of extra sanity checking
+options 	INVARIANT_SUPPORT	# Extra sanity checks of internal structures, required by INVARIANTS
+options 	WITNESS			# Enable checks to detect deadlocks and cycles
+options 	WITNESS_SKIPSPIN	# Don't run witness on spinlocks for speed
+options 	MALLOC_DEBUG_MAXZONES	# Separate malloc(9) zones
+
+# Disable minimum debugger support
+nooptions 	KDB_TRACE		# Print a stack trace for a panic.
--- /dev/null
+++ b/sys/conf/DEBIAN_NODEBUG
@@ -0,0 +1,18 @@
+nomakeoptions	DEBUG			# Build kernel with gdb(1) debug symbols
+nomakeoptions	WITH_CTF		# Run ctfconvert(1) for DTrace support
+
+nooptions	KDTRACE_FRAME		# Ensure frames are compiled in
+nooptions	KDTRACE_HOOKS		# Kernel DTrace hooks
+nooptions	DDB_CTF			# Kernel ELF linker loads CTF data
+
+nooptions 	DDB			# Support DDB.
+nooptions 	GDB			# Support remote GDB.
+nooptions 	DEADLKRES		# Enable the deadlock resolver
+nooptions 	INVARIANTS		# Enable calls of extra sanity checking
+nooptions 	INVARIANT_SUPPORT	# Extra sanity checks of internal structures, required by INVARIANTS
+nooptions 	WITNESS			# Enable checks to detect deadlocks and cycles
+nooptions 	WITNESS_SKIPSPIN	# Don't run witness on spinlocks for speed
+nooptions 	MALLOC_DEBUG_MAXZONES	# Separate malloc(9) zones
+
+# Enable minimum debugger support
+options 	KDB_TRACE		# Print a stack trace for a panic.
