--- a/sys/amd64/conf/GENERIC
+++ b/sys/amd64/conf/GENERIC
@@ -18,8 +18,8 @@
 #
 # $FreeBSD$
 
-cpu		HAMMER
-ident		GENERIC
+#cpu		HAMMER
+#ident		GENERIC
 
 makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
 makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
@@ -53,7 +53,7 @@
 options 	COMPAT_FREEBSD5		# Compatible with FreeBSD5
 options 	COMPAT_FREEBSD6		# Compatible with FreeBSD6
 options 	COMPAT_FREEBSD7		# Compatible with FreeBSD7
-options 	SCSI_DELAY=5000		# Delay (in ms) before probing SCSI
+options 	SCSI_DELAY=2000		# Delay (in ms) before probing SCSI
 options 	KTRACE			# ktrace(1) support
 options 	STACK			# stack(9) support
 options 	SYSVSHM			# SYSV-style shared memory
@@ -101,9 +101,6 @@
 device		acpi
 device		pci
 
-# Floppy drives
-device		fdc
-
 # ATA controllers
 device		ahci		# AHCI-compatible SATA controllers
 device		ata		# Legacy ATA/SATA controllers
@@ -201,9 +198,6 @@
 # Parallel port
 device		ppc
 device		ppbus		# Parallel port bus (required)
-device		lpt		# Printer
-device		ppi		# Parallel port interface device
-#device		vpo		# Requires scbus and da
 
 device		puc		# Multi I/O cards and multi-channel UARTs
 
@@ -301,10 +295,7 @@
 options 	RDRAND_RNG	# Intel Bull Mountain RNG
 device		ether		# Ethernet support
 device		vlan		# 802.1Q VLAN support
-device		tun		# Packet tunnel.
 device		md		# Memory "disks"
-device		gif		# IPv6 and IPv4 tunneling
-device		faith		# IPv6-to-IPv4 relaying (translation)
 device		firmware	# firmware assist module
 
 # The `bpf' device enables the Berkeley Packet Filter.
--- a/sys/i386/conf/GENERIC
+++ b/sys/i386/conf/GENERIC
@@ -18,10 +18,10 @@
 #
 # $FreeBSD$
 
-cpu		I486_CPU
-cpu		I586_CPU
-cpu		I686_CPU
-ident		GENERIC
+#cpu		I486_CPU
+#cpu		I586_CPU
+#cpu		I686_CPU
+#ident		GENERIC
 
 makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
 makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
@@ -46,6 +46,8 @@
 options 	MSDOSFS			# MSDOS Filesystem
 options 	CD9660			# ISO 9660 Filesystem
 options 	PROCFS			# Process filesystem (requires PSEUDOFS)
+options		LINPROCFS
+options		COMPAT_LINUX		# Needed by LINPROCFS
 options 	PSEUDOFS		# Pseudo-filesystem framework
 options 	GEOM_PART_GPT		# GUID Partition Tables.
 options 	GEOM_RAID		# Soft RAID functionality.
@@ -54,7 +56,7 @@
 options 	COMPAT_FREEBSD5		# Compatible with FreeBSD5
 options 	COMPAT_FREEBSD6		# Compatible with FreeBSD6
 options 	COMPAT_FREEBSD7		# Compatible with FreeBSD7
-options 	SCSI_DELAY=5000		# Delay (in ms) before probing SCSI
+options 	SCSI_DELAY=2000		# Delay (in ms) before probing SCSI
 options 	KTRACE			# ktrace(1) support
 options 	STACK			# stack(9) support
 options 	SYSVSHM			# SYSV-style shared memory
@@ -103,9 +105,6 @@
 device		eisa
 device		pci
 
-# Floppy drives
-device		fdc
-
 # ATA controllers
 device		ahci		# AHCI-compatible SATA controllers
 device		ata		# Legacy ATA/SATA controllers
@@ -212,9 +211,6 @@
 # Parallel port
 device		ppc
 device		ppbus		# Parallel port bus (required)
-device		lpt		# Printer
-device		ppi		# Parallel port interface device
-#device		vpo		# Requires scbus and da
 
 device		puc		# Multi I/O cards and multi-channel UARTs
 
@@ -315,10 +311,7 @@
 options 	RDRAND_RNG	# Intel Bull Mountain RNG
 device		ether		# Ethernet support
 device		vlan		# 802.1Q VLAN support
-device		tun		# Packet tunnel.
 device		md		# Memory "disks"
-device		gif		# IPv6 and IPv4 tunneling
-device		faith		# IPv6-to-IPv4 relaying (translation)
 device		firmware	# firmware assist module
 
 # The `bpf' device enables the Berkeley Packet Filter.
--- /dev/null
+++ b/sys/conf/DEBIAN
@@ -0,0 +1,33 @@
+# Common options to all Debian GNU/kFreeBSD kernels
+
+options		LINPROCFS
+options         LINSYSFS
+options		FDESCFS
+options		TMPFS
+options         QUOTA
+options		KTRACE
+
+# Network options
+device		carp
+device		pf
+device		pfsync
+device		pflog
+
+# pty(4) was removed from GENERIC in 10-CURRENT (see rev 233271),
+# but unfortunetely glibc doesn't support pts(4) yet (see #667448).
+device		pty
+
+# Toggle to enable/disable debug options.
+include		DEBIAN_DEBUG
+#include	DEBIAN_NODEBUG
+
+# Alternate queueing
+options		ALTQ
+options		ALTQ_CBQ	# Class Bases Queuing (CBQ)
+options		ALTQ_RED	# Random Early Detection (RED)
+options		ALTQ_RIO	# RED In/Out
+options		ALTQ_HFSC	# Hierarchical Packet Scheduler (HFSC)
+options		ALTQ_PRIQ	# Priority Queuing (PRIQ)
+
+# Disable binary blobs
+include		WITHOUT_SOURCELESS
--- /dev/null
+++ b/sys/conf/DEBIAN_DEBUG
@@ -0,0 +1,22 @@
+makeoptions	DEBUG=-g		# Build kernel with gdb(1) debug symbols
+
+## Not yet
+#makeoptions	WITH_CTF=1		# Run ctfconvert(1) for DTrace support
+nomakeoptions	WITH_CTF
+
+## Not yet (IMPORTANT: KDTRACE_FRAME is only for amd64!!)
+#options		KDTRACE_FRAME		# Ensure frames are compiled in
+#options		KDTRACE_HOOKS		# Kernel DTrace hooks
+#options		DDB_CTF			# Kernel ELF linker loads CTF data
+
+options 	DDB			# Support DDB.
+options 	GDB			# Support remote GDB.
+options 	DEADLKRES		# Enable the deadlock resolver
+options 	INVARIANTS		# Enable calls of extra sanity checking
+options 	INVARIANT_SUPPORT	# Extra sanity checks of internal structures, required by INVARIANTS
+options 	WITNESS			# Enable checks to detect deadlocks and cycles
+options 	WITNESS_SKIPSPIN	# Don't run witness on spinlocks for speed
+options 	MALLOC_DEBUG_MAXZONES	# Separate malloc(9) zones
+
+# Disable minimum debugger support
+nooptions 	KDB_TRACE		# Print a stack trace for a panic.
--- /dev/null
+++ b/sys/conf/DEBIAN_NODEBUG
@@ -0,0 +1,18 @@
+nomakeoptions	DEBUG			# Build kernel with gdb(1) debug symbols
+nomakeoptions	WITH_CTF		# Run ctfconvert(1) for DTrace support
+
+nooptions	KDTRACE_FRAME		# Ensure frames are compiled in
+nooptions	KDTRACE_HOOKS		# Kernel DTrace hooks
+nooptions	DDB_CTF			# Kernel ELF linker loads CTF data
+
+nooptions 	DDB			# Support DDB.
+nooptions 	GDB			# Support remote GDB.
+nooptions 	DEADLKRES		# Enable the deadlock resolver
+nooptions 	INVARIANTS		# Enable calls of extra sanity checking
+nooptions 	INVARIANT_SUPPORT	# Extra sanity checks of internal structures, required by INVARIANTS
+nooptions 	WITNESS			# Enable checks to detect deadlocks and cycles
+nooptions 	WITNESS_SKIPSPIN	# Don't run witness on spinlocks for speed
+nooptions 	MALLOC_DEBUG_MAXZONES	# Separate malloc(9) zones
+
+# Enable minimum debugger support
+options 	KDB_TRACE		# Print a stack trace for a panic.
