--- a/sys/conf/DEBIAN
+++ b/sys/conf/DEBIAN
@@ -17,28 +17,8 @@
 # but unfortunetely glibc doesn't support pts(4) yet (see #667448).
 device		pty
 
-# Full debugger support is disabled by upstream when a branch is close to
-# release.  Unfortunately this can trigger the appearance of heisenbugs
-# which were previously hidden (we've been hit by this already, see #651624).
-#
-# This disables full debugger support in advance, so that the build
-# configuration which is going to be used for release is tested during the
-# whole release cycle.  If you need a debug kernel, simply remove/comment
-# the following lines:
-
-nooption 	DDB			# Support DDB.
-nooption 	GDB			# Support remote GDB.
-nooption 	DEADLKRES		# Enable the deadlock resolver
-nooption 	INVARIANTS		# Enable calls of extra sanity checking
-nooption 	INVARIANT_SUPPORT	# Extra sanity checks of internal structures, required by INVARIANTS
-nooption 	WITNESS			# Enable checks to detect deadlocks and cycles
-nooption 	WITNESS_SKIPSPIN	# Don't run witness on spinlocks for speed
-nooption 	MALLOC_DEBUG_MAXZONES	# Separate malloc(9) zones
-
-# Usually when full debugger is enabled, minimum debugger support is disabled.
-# Bring it back:
-
-options 	KDB_TRACE		# Print a stack trace for a panic.
+# Build with GDB debugging symbols.
+makeoptions	DEBUG=-g
 
 # Alternate queueing
 options		ALTQ
